<a href="https://colab.research.google.com/github/HammadN98/previsao_de_vendas/blob/main/previsao_das_vendas_lojas_Rossmann.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 0. Importacoes

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [2]:
store = pd.read_csv("https://raw.githubusercontent.com/HammadN98/previsao_de_vendas/refs/heads/main/rossmann-store-sales/store.csv")
store.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

In [3]:
train = pd.read_csv("https://raw.githubusercontent.com/HammadN98/previsao_de_vendas/refs/heads/main/rossmann-store-sales/train.csv")
train.head()

<ipython-input-3-24519896a09f>:1: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  train = pd.read_csv("https://raw.githubusercontent.com/HammadN98/previsao_de_vendas/refs/heads/main/rossmann-store-sales/train.csv")


Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [4]:
test = pd.read_csv("https://raw.githubusercontent.com/HammadN98/previsao_de_vendas/refs/heads/main/rossmann-store-sales/test.csv")
test.head()

Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4  2015-09-17   1.0      1            0              0
1   2      3          4  2015-09-17   1.0      1            0              0
2   3      7          4  2015-09-17   1.0      1            0              0
3   4      8          4  2015-09-17   1.0      1            0              0
4   5      9          4  2015-09-17   1.0      1            0              0

juntar os datasets para garantir ter todas as variaveis disponiveis

In [5]:
#Dimensoes pre-merge
store.shape, train.shape, test.shape

((1115, 10), (1017209, 9), (41088, 8))

In [6]:
train = train.merge(store, on='Store', how='left')
train.head()


Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday StoreType Assortment  CompetitionDistance  \
0              1         c          a               1270.0   
1              1         a          a                570.0   
2              1         a          a              14130.0   
3              1         c          c                620.0   
4              1         a          a              29910.0   

   CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
0                        9.0                    2008.0       0   
1                       11.0                    2007.0       1   
2                       12.0                    2006.0       1   
3                        9.0                    2009.0       0   
4                        4.0                    2015.0       0   

   Promo2SinceWeek  Promo2SinceYear    PromoInterval  
0              NaN              NaN              NaN  
1             13.0           2010.0  Jan,Apr,Jul,Oct  
2             14.0           2011.0  Jan,Apr,Jul,Oct  
3              NaN              NaN              NaN  
4              NaN              NaN              NaN

In [7]:
test = test.merge(store, on='Store', how='left')
test.head()

Id  Store  DayOfWeek        Date  Open  Promo StateHoliday  SchoolHoliday  \
0   1      1          4  2015-09-17   1.0      1            0              0   
1   2      3          4  2015-09-17   1.0      1            0              0   
2   3      7          4  2015-09-17   1.0      1            0              0   
3   4      8          4  2015-09-17   1.0      1            0              0   
4   5      9          4  2015-09-17   1.0      1            0              0   

  StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0         c          a               1270.0                        9.0   
1         a          a              14130.0                       12.0   
2         a          c              24000.0                        4.0   
3         a          a               7520.0                       10.0   
4         a          c               2030.0                        8.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2006.0       1             14.0           2011.0   
2                    2013.0       0              NaN              NaN   
3                    2014.0       0              NaN              NaN   
4                    2000.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2              NaN  
3              NaN  
4              NaN

In [8]:
#Dimensoes pOS-merge
train.shape, test.shape


((1017209, 18), (41088, 17))

## 1. Analise Exploratoria

### 1.1 Valores ausentes

In [9]:
#Valores ausentes por coluna no train
train.isnull().sum()

Store                             0
DayOfWeek                         0
Date                              0
Sales                             0
Customers                         0
Open                              0
Promo                             0
StateHoliday                      0
SchoolHoliday                     0
StoreType                         0
Assortment                        0
CompetitionDistance            2642
CompetitionOpenSinceMonth    323348
CompetitionOpenSinceYear     323348
Promo2                            0
Promo2SinceWeek              508031
Promo2SinceYear              508031
PromoInterval                508031
dtype: int64

In [10]:
#Valores ausentes por coluna no test
test.isnull().sum()

Id                               0
Store                            0
DayOfWeek                        0
Date                             0
Open                            11
Promo                            0
StateHoliday                     0
SchoolHoliday                    0
StoreType                        0
Assortment                       0
CompetitionDistance             96
CompetitionOpenSinceMonth    15216
CompetitionOpenSinceYear     15216
Promo2                           0
Promo2SinceWeek              17232
Promo2SinceYear              17232
PromoInterval                17232
dtype: int64

A coluna usada como base para o merge ficou com 0 NaN

In [11]:
#PErcentual de valores ausentes por coluna
print(train.isnull().mean() * 100)
print(test.isnull().mean() * 100)


Store                         0.000000
DayOfWeek                     0.000000
Date                          0.000000
Sales                         0.000000
Customers                     0.000000
Open                          0.000000
Promo                         0.000000
StateHoliday                  0.000000
SchoolHoliday                 0.000000
StoreType                     0.000000
Assortment                    0.000000
CompetitionDistance           0.259730
CompetitionOpenSinceMonth    31.787764
CompetitionOpenSinceYear     31.787764
Promo2                        0.000000
Promo2SinceWeek              49.943620
Promo2SinceYear              49.943620
PromoInterval                49.943620
dtype: float64
Id                            0.000000
Store                         0.000000
DayOfWeek                     0.000000
Date                          0.000000
Open                          0.026772
Promo                         0.000000
StateHoliday                  0.000000
SchoolHoli

OS valores faltantes serao substituidos por:

CompetitionDistance -> Mediana

Open -> 0 / A loja devia estar fechada

CompetitionOpenSinceMonth, CompetitionOpenSinceYear -> 0 / indica que nao tem competidores proximos

Promo2SinceWeek, Promo2SinceYear, PromoInterval -> 0 / A loja nao participas das promocoes

In [12]:
train['CompetitionDistance']

0           1270.0
1            570.0
2          14130.0
3            620.0
4          29910.0
            ...   
1017204     1900.0
1017205     1880.0
1017206     9260.0
1017207      870.0
1017208     5350.0
Name: CompetitionDistance, Length: 1017209, dtype: float64

In [13]:
# NaN -> MEdiana
median_value_train = train['CompetitionDistance'].median()
train['CompetitionDistance'] = train['CompetitionDistance'].fillna(median_value_train)
median_value_test = test['CompetitionDistance'].median()
test['CompetitionDistance'] = test['CompetitionDistance'].fillna(median_value_test)

# NaN -> 0
train['CompetitionOpenSinceMonth'] = train['CompetitionOpenSinceMonth'].fillna(0)
test['CompetitionOpenSinceMonth'] = test['CompetitionOpenSinceMonth'].fillna(0)

train['CompetitionOpenSinceYear'] = train['CompetitionOpenSinceYear'].fillna(0)
test['CompetitionOpenSinceYear'] = test['CompetitionOpenSinceYear'].fillna(0)

train['Promo2SinceWeek'] = train['Promo2SinceWeek'].fillna(0)
test['Promo2SinceWeek'] = test['Promo2SinceWeek'].fillna(0)

train['Promo2SinceYear'] = train['Promo2SinceYear'].fillna(0)
test['Promo2SinceYear'] = test['Promo2SinceYear'].fillna(0)

# NaN -> 'None'
train['PromoInterval'] = train['PromoInterval'].fillna('None')
test['PromoInterval'] = test['PromoInterval'].fillna('None')

test['Open'] = test['Open'].fillna(0)

In [14]:
print("Valores ausentes após o tratamento (train):")
print(train.isnull().sum())

Valores ausentes após o tratamento (train):
Store                        0
DayOfWeek                    0
Date                         0
Sales                        0
Customers                    0
Open                         0
Promo                        0
StateHoliday                 0
SchoolHoliday                0
StoreType                    0
Assortment                   0
CompetitionDistance          0
CompetitionOpenSinceMonth    0
CompetitionOpenSinceYear     0
Promo2                       0
Promo2SinceWeek              0
Promo2SinceYear              0
PromoInterval                0
dtype: int64


In [15]:
print("\nValores ausentes após o tratamento (test):")
print(test.isnull().sum())


Valores ausentes após o tratamento (test):
Id                           0
Store                        0
DayOfWeek                    0
Date                         0
Open                         0
Promo                        0
StateHoliday                 0
SchoolHoliday                0
StoreType                    0
Assortment                   0
CompetitionDistance          0
CompetitionOpenSinceMonth    0
CompetitionOpenSinceYear     0
Promo2                       0
Promo2SinceWeek              0
Promo2SinceYear              0
PromoInterval                0
dtype: int64


Dados faltantes tratados

## 2. Analise Descritiva

In [16]:
train.describe()

Store     DayOfWeek         Sales     Customers          Open  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
mean   5.584297e+02  3.998341e+00  5.773819e+03  6.331459e+02  8.301067e-01   
std    3.219087e+02  1.997391e+00  3.849926e+03  4.644117e+02  3.755392e-01   
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.800000e+02  2.000000e+00  3.727000e+03  4.050000e+02  1.000000e+00   
50%    5.580000e+02  4.000000e+00  5.744000e+03  6.090000e+02  1.000000e+00   
75%    8.380000e+02  6.000000e+00  7.856000e+03  8.370000e+02  1.000000e+00   
max    1.115000e+03  7.000000e+00  4.155100e+04  7.388000e+03  1.000000e+00   

              Promo  SchoolHoliday  CompetitionDistance  \
count  1.017209e+06   1.017209e+06         1.017209e+06   
mean   3.815145e-01   1.786467e-01         5.422034e+03   
std    4.857586e-01   3.830564e-01         7.706913e+03   
min    0.000000e+00   0.000000e+00         2.000000e+01   
25%    0.000000e+00   0.000000e+00         7.100000e+02   
50%    0.000000e+00   0.000000e+00         2.330000e+03   
75%    1.000000e+00   0.000000e+00         6.880000e+03   
max    1.000000e+00   1.000000e+00         7.586000e+04   

       CompetitionOpenSinceMonth  CompetitionOpenSinceYear        Promo2  \
count               1.017209e+06              1.017209e+06  1.017209e+06   
mean                4.926878e+00              1.370173e+03  5.005638e-01   
std                 4.283543e+00              9.353634e+02  4.999999e-01   
min                 0.000000e+00              0.000000e+00  0.000000e+00   
25%                 0.000000e+00              0.000000e+00  0.000000e+00   
50%                 4.000000e+00              2.006000e+03  1.000000e+00   
75%                 9.000000e+00              2.011000e+03  1.000000e+00   
max                 1.200000e+01              2.015000e+03  1.000000e+00   

       Promo2SinceWeek  Promo2SinceYear  
count     1.017209e+06     1.017209e+06  
mean      1.164767e+01     1.007011e+03  
std       1.532393e+01     1.005877e+03  
min       0.000000e+00     0.000000e+00  
25%       0.000000e+00     0.000000e+00  
50%       1.000000e+00     2.009000e+03  
75%       2.200000e+01     2.012000e+03  
max       5.000000e+01     2.015000e+03

#### Insights:

Store: Vai de 1 a 1115, indicando 1115 loja unicas.

DayOfWeek: VAi de 1 (Segunda) a 7 (Domingo).

Sales: VEndas minimas = 0, possivelvente a loja estava fechada, a media fica em 5773.82, com um desvio padrao de 3849.93 o que indica uma altissima variacao. E tando o maximo em 41,551.

Custumers: MIN -> 0(Loja fechada possivelmente). Media -> 633. Dp -> 464, indicando alta variacao.  MAX -> 7388.

Promo: A media indica que 38% dos dados tem promocoes em atividade.

Promo2: A media indica que +50% nao participa das promocoes 2

CompetitionDistance: MIN -> 2km indica proximidade de concorrentes, Media -> 5,4km. MAX -> 75km, lojas mais isoladas.

CompetitionOpenSinceYear -> OS valores 0, puxaram a media para 1370. Indicando que a escolha da codificacao nao foi efetiva. Retomar a esse ponto.

Open: 83% da lojas abertas

Holiday(School e State): Grande maioria 0, o que era esperado.






### 2. 1 Validando `CompetitionOpenSinceYear`

In [17]:
print("Valores únicos de CompetitionOpenSinceYear:")
print(train['CompetitionOpenSinceYear'].value_counts().sort_index())

print("\nAno mínimo e máximo:")
print("Mínimo:", train['CompetitionOpenSinceYear'].min())
print("Máximo:", train['CompetitionOpenSinceYear'].max())

Valores únicos de CompetitionOpenSinceYear:
CompetitionOpenSinceYear
0.0       323348
1900.0       758
1961.0       942
1990.0      4710
1994.0      1884
1995.0      1700
1998.0       942
1999.0      7352
2000.0      9236
2001.0     14704
2002.0     24882
2003.0     16794
2004.0     19804
2005.0     56564
2006.0     42802
2007.0     43744
2008.0     48476
2009.0     49396
2010.0     51258
2011.0     49396
2012.0     74299
2013.0     75426
2014.0     63732
2015.0     35060
Name: count, dtype: int64

Ano mínimo e máximo:
Mínimo: 0.0
Máximo: 2015.0


os valores em zero, serao substituidos pela mediana, e sera criada uma nova categoria, binaria, para indicar quais sao as lojas que nao possuiam informacao.

In [18]:
#Substituir 0 -> NaN
train['CompetitionOpenSinceYear'] = train['CompetitionOpenSinceYear'].replace(0, np.nan)

In [19]:
#CRiando a nova coluna
train['CompetitionMissing'] = train['CompetitionOpenSinceYear'].isna().astype(int)

In [20]:
#Ausentes -> mediana (2010)
mediana_year = train['CompetitionOpenSinceYear'].median()
train['CompetitionOpenSinceYear'] = train['CompetitionOpenSinceYear'].fillna(mediana_year)

In [21]:
train[['CompetitionOpenSinceYear', 'CompetitionMissing']]

CompetitionOpenSinceYear  CompetitionMissing
0                          2008.0                   0
1                          2007.0                   0
2                          2006.0                   0
3                          2009.0                   0
4                          2015.0                   0
...                           ...                 ...
1017204                    2014.0                   0
1017205                    2006.0                   0
1017206                    2010.0                   1
1017207                    2010.0                   1
1017208                    2010.0                   1

[1017209 rows x 2 columns]

In [22]:
#Para os dados de teste tbm

#Substituir 0 -> NaN
test['CompetitionOpenSinceYear'] = test['CompetitionOpenSinceYear'].replace(0, np.nan)

#CRiando a nova coluna
test['CompetitionMissing'] =test['CompetitionOpenSinceYear'].isna().astype(int)

#Ausentes -> mediana
mediana_year_test = test['CompetitionOpenSinceYear'].median()
test['CompetitionOpenSinceYear'] = test['CompetitionOpenSinceYear'].fillna(mediana_year_test)

test[['CompetitionOpenSinceYear', 'CompetitionMissing']]

CompetitionOpenSinceYear  CompetitionMissing
0                        2008.0                   0
1                        2006.0                   0
2                        2013.0                   0
3                        2014.0                   0
4                        2000.0                   0
...                         ...                 ...
41083                    2014.0                   0
41084                    2006.0                   0
41085                    2010.0                   1
41086                    2010.0                   1
41087                    2010.0                   1

[41088 rows x 2 columns]

### 2.2 Analisando as variaveis categoricas

In [23]:
cat_columns = train.select_dtypes(include='object').columns
cat_columns

Index(['Date', 'StateHoliday', 'StoreType', 'Assortment', 'PromoInterval'], dtype='object')

In [29]:
for categorica in cat_columns:
    print(f"{train[categorica].value_counts()}\n")

Date
2015-07-31    1115
2013-11-06    1115
2013-11-18    1115
2013-11-17    1115
2013-11-16    1115
              ... 
2014-10-28     935
2014-10-27     935
2014-10-26     935
2014-10-25     935
2014-12-08     935
Name: count, Length: 942, dtype: int64

StateHoliday
0    855087
0    131072
a     20260
b      6690
c      4100
Name: count, dtype: int64

StoreType
a    551627
d    312912
c    136840
b     15830
Name: count, dtype: int64

Assortment
a    537445
c    471470
b      8294
Name: count, dtype: int64

PromoInterval
None                508031
Jan,Apr,Jul,Oct     293122
Feb,May,Aug,Nov     118596
Mar,Jun,Sept,Dec     97460
Name: count, dtype: int64



StateHoliday: Possui dois zeros. PRovalemnete um numerico e outro string
StoreType: ok
Assoretment: ok
PromoInterval: ok
As variaveis categoricas quando estiverem ok, ainda precisam ser transformadas para variaveis numericas.


#### 2.2.2 REsolvendo as inconsistencias em StateHoliday

In [31]:
train['StateHoliday'] = train['StateHoliday'].astype(str).replace('0.0', '0')
test['StateHoliday'] = test['StateHoliday'].astype(str).replace('0.0', '0')


In [36]:
train['StateHoliday'].value_counts()

StateHoliday
0    986159
a     20260
b      6690
c      4100
Name: count, dtype: int64

### 2.3 Transformando as Variaveis Categoricas

Sera utilizado o One-Hot

In [37]:
train_one_hot = pd.get_dummies(train, columns=['StateHoliday'], prefix='StateHoliday', drop_first=True)
train_one_hot = pd.get_dummies(train, columns=['PromoInterval'], prefix='PromoInterval', drop_first=True)
train_one_hot = pd.get_dummies(train, columns=['StoreType'], prefix='StoreType', drop_first=True)
train_one_hot = pd.get_dummies(train, columns=['Assortment'], prefix='Assortment', drop_first=True)


In [38]:
test_one_hot = pd.get_dummies(test, columns=['StateHoliday'], prefix='StateHoliday', drop_first=True)
test_one_hot = pd.get_dummies(test, columns=['PromoInterval'], prefix='PromoInterval', drop_first=True)
test_one_hot = pd.get_dummies(test, columns=['StoreType'], prefix='StoreType', drop_first=True)
test_one_hot = pd.get_dummies(test, columns=['Assortment'], prefix='Assortment', drop_first=True)

In [40]:
train_one_hot.columns

Index(['Store', 'DayOfWeek', 'Date', 'Sales', 'Customers', 'Open', 'Promo',
       'StateHoliday', 'SchoolHoliday', 'StoreType', 'CompetitionDistance',
       'CompetitionOpenSinceMonth', 'CompetitionOpenSinceYear', 'Promo2',
       'Promo2SinceWeek', 'Promo2SinceYear', 'PromoInterval',
       'CompetitionMissing', 'Assortment_b', 'Assortment_c'],
      dtype='object')

In [41]:
test_one_hot.columns

Index(['Id', 'Store', 'DayOfWeek', 'Date', 'Open', 'Promo', 'StateHoliday',
       'SchoolHoliday', 'StoreType', 'CompetitionDistance',
       'CompetitionOpenSinceMonth', 'CompetitionOpenSinceYear', 'Promo2',
       'Promo2SinceWeek', 'Promo2SinceYear', 'PromoInterval',
       'CompetitionMissing', 'Assortment_b', 'Assortment_c'],
      dtype='object')